satellite.go

```diff
@@ -0,0 +1,214 @@
+					DurationMs string `json:"duration_ms"`
+				} `json:"pulp3"`
+				Pulp3Content struct {
+					Status     string `json:"status"`
+					DurationMs string `json:"duration_ms"`
+				} `json:"pulp3_content"`
+			} `json:"services"`
+			Status string `json:"status"`
+		} `json:"katello"`
+	} `json:"results"`
+}
+
+// ConfigureSatelliteResult is structure holding information about results
+// of configuring host from Satellite server. The result could be printed
+// in machine-readable format.
+type ConfigureSatelliteResult struct {
+	Hostname                 string `json:"hostname"`
+	HostnameError            string `json:"hostname_error,omitempty"`
+	IsServerSatellite        bool   `json:"is_server_satellite_running"`
+	SatelliteServerHostname  string `json:"satellite_server_hostname"`
+	SatelliteServerScriptUrl string `json:"satellite_server_script_url"`
+	HostConfigured           bool   `json:"host_configured"`
+	format                   string
+}
+
+// Error implement error interface for structure ConfigureSatelliteResult
+func (configureSatelliteResult ConfigureSatelliteResult) Error() string {
+	var result string
+	switch configureSatelliteResult.format {
+	case "json":
+		data, err := json.MarshalIndent(configureSatelliteResult, "", "    ")
+		if err != nil {
+			return err.Error()
+		}
+		result = string(data)
+	case "":
+		break
+	default:
+		result = "error: unsupported document format: " + configureSatelliteResult.format
+	}
+	return result
+}
+
+// pingSatelliteServer tries to ping Satellite server to be sure that user
+// tries to reach get and run bootstrap script from Satellite server.
+// We use following endpoint that is available for unauthenticated users:
+// https://satellite.sat.engineering.redhat.com/apidoc/v2/ping/ping.en.html
+// It is not necessary to try to reach this endpoint, but download and run
+// some bash script downloaded from the internet. It is risky action. For
+// this reason we do this extra step to minimize risk that some non-intentional
+// code is downloaded and run as root user.
+//
+// In the future we could use following endpoint to determine version of
+// Satellite (Katello) and
+// https://satellite.sat.engineering.redhat.com/apidoc/v2/ping/server_status.en.html
+func pingSatelliteServer(satelliteScriptUrl *url.URL, s *spinner.Spinner) error {
+	if s != nil {
+		s.Suffix = fmt.Sprintf(" Connecting Satellite server: %v", satelliteScriptUrl.Host)
+	}
+
+	// Copy URL from struct URL and modify scheme, path and query
+	satellitePingUrl := *satelliteScriptUrl
+	satellitePingUrl.Scheme = "https"
+	satellitePingUrl.Path = "/api/ping"
+	satellitePingUrl.RawQuery = ""
+
+	// We have to use insecure HTTPs connection, because most of the customers use
+	// self-signed certificates
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: true,
+	}
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.TLSClientConfig = &tlsConfig
+	client := &http.Client{Transport: transport}
```

Comment:
Is this same client (with its TLS configuration) not required to download the
script in `downloadSatelliteScript`?

Reply:
It seems that it is possible to access `/pub` using (http and https). It is possible to access `/api/ping` using only https. In theory we can try to use one client using only one https connection.

Reply:
Yea, I think we can consolidate the HTTP clients into a specialized one for
performing Satellite REST requests.